aries_cloudagent/wallet_handler/handler.py

```diff
@@ -228,6 +248,33 @@ def get_paths(self, wallet: str) -> []:
 
         return handles
 
+    async def get_wallet_for_msg(self, body: bytes) -> [str]:
+        """
+        Parses an inbound message for recipient keys and returns the wallets
+        associated to keys.
+
+        Args:
+            body: Inbound raw message
+
+        Raises:
+            KeyNotFoundError: if given key does not belong to handled_keys
+        """
+        msg = json.loads(body)
+        protected = json.loads(b64decode(msg['protected']))
+        recipients = protected['recipients']
+        wallet_ids = []
+        # Check each recipient public key (in `kid`) if agent handles a wallet
+        # associated to that key.
+        for recipient in recipients:
+            kid = recipient['header']['kid']
+            try:
+                wallet_id = await self.get_wallet_for_key(kid)
+            except KeyNotFoundError:
+                wallet_id = None
+            wallet_ids.append(wallet_id)
+
+        return wallet_ids
+
     async def get_wallet_for_path(self, path: str) -> str:
```

Comment:
To associate a wallet to the inbound message we parse the body and check the key send in `kid` with the mapping stored in the handler instead of having an inbound path specific for that connection which than is  used as mapping to reference wallet.

Reply:
Does the other agent need to do anything differently for this to work?

Reply:
No the other agent doesn't need to know if it is a multitenancy or standard agent. It just uses the informtion defined in standard [Aries RFC 0019: Encryption Envelope](https://github.com/hyperledger/aries-rfcs/tree/master/features/0019-encryption-envelope)